RN navigation (#62)
infinum/frontend-handbook: React Native/Navigation.md

> Routing in mobile applications presents more then mere navigation between routes (screens). It consists of functionalities and styles which are divided across platforms. Generally, routing systems inside native applications has its own look and feel aswell as well defined UX which differs from android and iOS platforms.

Reviewer: `has its own look and feel aswell as well defined UX which differs` should probably be
`has its own look and feel as well defined UX which differs`

In order to achieve this “native feel”, RN community implemented a [React Navigation](https://reactnavigation.org/) library. It’s main focus presents “native-look-and-feel” while keeping the performance and extensibility.

### Official documentation

> [Official documentation](https://reactnavigation.org/docs/getting-started) has a lot of informations and contains more then few code examples of all important parts.

Reviewer 2: ```suggestion
[Official documentation](https://reactnavigation.org/docs/getting-started) has a lot of information and contains more than a few code examples of all important parts.
```

> Make sure to get faimiliar with it before continue to read further.

Reviewer 2: ```suggestion
Make sure to get familiar with it before continuing to read further.
```

### 1. Navigation file structure

> In order to keep your navigation maintanable and easy to upgrade  here are some guidelines which rely on library main concepts: ***Stack***, ***Tab***, ***Drawer*** and ***Parameters***.

Reviewer 2: ```suggestion
In order to keep your navigation maintainable and easy to upgrade, here are some guidelines which rely on library main concepts: ***Stack***, ***Tab***, ***Drawer*** and ***Parameters***.
```

 - navigation/

> 	  -- stacks/

Reviewer 2: ```suggestion
	 - stacks/
```

> 	  -- tabs/

Reviewer 2: ```suggestion
	 - tabs/
```

> 	  -- drawer/

Reviewer 2: ```suggestion
	 - drawer/
```

> 	  -- params/

Reviewer 2: ```suggestion
	 - params/
```

> 	  -- Navigator.tsx

Reviewer 2: ```suggestion
	 - Navigator.tsx
```

#### 1.1 `Navigator.tsx`
 `Navigator.tsx` is a React functional component which defines navigation entry point. It should contain navigation structure wrapped inside `NavigationContainer`:

```typescript
const  Navigator:  React.FC<INavigatorProps>  =  (props)  =>  {
	return  (
		<NavigationContainer>
			// NAVIGATION STRUCTURE
		</NavigationContainer>
	);
};
```

#### 1.2 STACKS
All navigation stacks should be wrapped as functional components inside `navigation/stacks`.
For example, most common application structure consits of authentication (*login*, *register*) and main (*home*, etc..) workflow.

> In React Navigation terms this means that you should split your navigation stacks between those 2 flows: `AuthenticationStack` and `HomeStack`. Even though, we could use only one stack with all screens, this is a bad practise in mobile development and makes the app harder to maintain as your application grows.

Reviewer 2: ```suggestion
In React Navigation terms this means that you should split your navigation stacks between those 2 flows: `AuthenticationStack` and `HomeStack`. Even though we could use only one stack with all screens, this is a bad practice in mobile development and makes the app harder to maintain as your application grows.
```

Using example above this `Navigator.tsx` would look like:

 ```typescript
 const RootStack = createStackNavigator();

const Navigator: React.FC<INavigatorProps> = (props) => {
	return (
		<NavigationContainer>
			<RootStack.Navigator>
				<RootStack.Screen component={AuthenticationStack} name="AuthenticationStack" />
				<RootStack.Screen component={HomeStack} name="HomeStack" />
			</RootStack.Navigator>
		</NavigationContainer>
	);
};
```

#### 1.3 TABS
Usually, applications have one navigation tab which should be in `navigation/tabs`. When using tab navigation, each `Tab.Screen` component should target a stack with related screens.

 ```typescript
 const Tab = createTabNavigator();

const TabNavigator: React.FC<ITabNavigatorProps> = (props) => {
	return (
		<Tab.Navigator>
			<Tab.Screen component={AuthenticationStack} name="AuthenticationStack" />
			<Tab.Screen component={HomeStack} name="HomeStack" />
		</Tab.Navigator>
	);
};
```

#### 1.4 Drawer

> Usually, applications have one navigation tab which should be in `navigation/drawer`. When using tab navigation, each `Tab.Screen` component should target a stack with related screens.

Reviewer 3: Should be Drawer.Screen ?

Author: Good catch!

 ```javascript
const Drawer = createDrawerNavigator();

const DrawerNavigator: React.FC<IDrawerNavigatorProps> = (props) => {
	return (
		<Drawer.Navigator>
			<Drawer.Screen component={AuthenticationStack} name="AuthenticationStack" />
			<Drawer.Screen component={HomeStack} name="HomeStack" />
		</Drawer.Navigator>
	);
};
```

#### 1.4 Parameters (Typescript)
To fully use the power of Typescript `navigation/params` folder should contain types and interfaces for all stacks used inside application. Since React Navigation handles `Tab` and `Drawer` navigations in background there is no need for any type definitions.
Each stack should have its parameters defined in separate file. Parameters file consists of 3 main parts:

 1. List of screens used inside of stack with related `route` params defined as `type`.
 2. Screen `route` parameters defined as `interface`
 3. Stack screen props defined as `type`

Example:
 1.

 ```javascript
export type HomeStackParamList = {
	Home: IHomeRouteProps;
	// ... other HomeStack screens
};
```
2.
 ```javascript
export interface IHomeRouteProps {
	propName: propValue;
	...
};
```
3.

>  ```javascript

Author: I agree.

> export type HomeStackScreenProps<T extends keyof HomeStackParamList = {

Reviewer 3: Needs to be with closing tag:

HomeStackScreenProps<T extends keyof HomeStackParamList **>**
